sidecar/src/service/session_info.rs

```diff
@@ -0,0 +1,249 @@
+    #[cfg(feature = "tracing")]
+    pub session_id: String,
+}
+
+impl SessionInfo {
+    /// Returns the `RuntimeInfo` for a given runtime ID.
+    ///
+    /// If the runtime does not exist, it creates a new one and returns it.
+    ///
+    /// # Arguments
+    ///
+    /// * `runtime_id` - The ID of the runtime.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use datadog_sidecar::service::SessionInfo;
+    ///
+    /// #[tokio::main]
+    /// async fn main() {
+    ///     let session_info = SessionInfo::default();
+    ///     let runtime_info = session_info.get_runtime(&"runtime1".to_string());
+    /// }
+    /// ```
+    // DEV-TODO: This function should likely either be refactored or have its name changed as its
+    // performing a get or insert operation.
+    pub fn get_runtime(&self, runtime_id: &String) -> crate::interface::RuntimeInfo {
+        let mut runtimes = self.lock_runtimes();
+        match runtimes.get(runtime_id) {
+            Some(runtime) => runtime.clone(),
+            None => {
+                let mut runtime = crate::interface::RuntimeInfo::default();
+                runtimes.insert(runtime_id.clone(), runtime.clone());
+                #[cfg(feature = "tracing")]
+                if enabled!(Level::INFO) {
+                    runtime.instance_id = InstanceId {
+                        session_id: self.session_id.clone(),
+                        runtime_id: runtime_id.clone(),
+                    };
+                    info!(
+                        "Registering runtime_id {} for session {}",
+                        runtime_id, self.session_id
+                    );
+                }
+                runtime
+            }
+        }
+    }
+
+    /// Shuts down all runtimes in the session.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use datadog_sidecar::service::SessionInfo;
+    ///
+    /// #[tokio::main]
+    /// async fn main() {
+    ///     let session_info = SessionInfo::default();
+    ///     session_info.shutdown().await;
+    /// }
+    /// ```
+    pub async fn shutdown(&self) {
+        let runtimes: Vec<crate::interface::RuntimeInfo> = self
+            .lock_runtimes()
+            .drain()
+            .map(|(_, instance)| instance)
+            .collect();
+
+        let runtimes_shutting_down: Vec<_> = runtimes
+            .into_iter()
+            .map(|rt| tokio::spawn(async move { rt.shutdown().await }))
+            .collect();
+
+        future::join_all(runtimes_shutting_down).await;
+    }
+    /// Shuts down all running instances in the session.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use datadog_sidecar::service::SessionInfo;
+    ///
+    /// #[tokio::main]
+    /// async fn main() {
+    ///     let session_info = SessionInfo::default();
+    ///     session_info.shutdown_running_instances().await;
+    /// }
+    /// ```
+    pub async fn shutdown_running_instances(&self) {
```

Comment:
Calling this function out for review as it has not only moved, but changed. The previous implementation was cloning the runtimes and never actually shutting them down. This implementation drains the runtimes hashmap and then shuts them down. 

I doubt this is the optimal implementation, but it at least works as advertised now.